include/kokkos-utils/callbacks/TimerListener.hpp

```diff
@@ -0,0 +1,109 @@
+#ifndef KOKKOS_UTILS_CALLBACKS_TIMERLISTENER_HPP
+#define KOKKOS_UTILS_CALLBACKS_TIMERLISTENER_HPP
+
+#include "Kokkos_NumericTraits.hpp"
+
+#include "kokkos-utils/callbacks/Listener.hpp"
+#include "kokkos-utils/callbacks/Matcher.hpp"
+#include "kokkos-utils/concepts/ExecutionSpace.hpp"
+
+namespace Kokkos::utils::callbacks
+{
+
+template <typename TimerType>
+concept Startable = requires (TimerType& timer) {
+    { timer.start() } -> std::same_as<void>;
+};
+
+template <typename TimerType, typename EventType>
+concept StartableWithEvent = requires (TimerType& timer, const EventType& event) {
+    { timer.start(event) } -> std::same_as<void>;
+};
+
+template <typename TimerType>
+concept Stoppable = requires (TimerType& timer) {
+    { timer.stop() } -> std::same_as<void>;
+};
+
+template <typename TimerType, typename EventType>
+concept StoppableWithEvent = requires (TimerType& timer, const EventType& event) {
+    { timer.stop(event) } -> std::same_as<void>;
+};
+
+/**
+ * @brief Listener that starts a timer when a begin event is received and stops it
+ *        when a corresponding end event is received.
+ *
+ * The begin event is selected by requiring it to be a match for @p BeginEndMatcherType
+ * while no begin event has been matched before. The end event is selected by requiring
+ * it to be a match for @p BeginEndMatcherType after a begin event has been matched.
+ *
+ * The listener can be reset by using the member function @ref reset. After calling
+ * @ref reset, a new begin event can be matched.
+ */
+template <
+    Matcher BeginEndMatcherType,
+    typename TimerType
+>
+class TimerListener
+{
+public:
+    //! @todo Check if this constructor is needed.
+    template <typename T, typename U = TimerType>
+    TimerListener(T&& matcher_, U&& timer_ = TimerType{})
+      : matcher{std::forward<T>(matcher_)}, timer{std::forward<U>(timer_)} {}
+
+    template <Event EventType> requires MatcherFor<BeginEndMatcherType, EventType>
+    void operator()(const EventType& event)
+    {
+        if(closed()) return;
+
+        const bool matching = matcher(event);
+
+        // Start when the event matches while no begin event has been matched before.
+        if(!this->matched_begin && matching)
+        {
+            this->matched_begin = true;
+
+            if constexpr (Startable<TimerType>) {
+                this->timer.start();
+            } else if constexpr (StartableWithEvent<TimerType, EventType>) {
+                this->timer.start(event);
+            }
+        }
+        // Stop when the event matches after a begin event has been matched.
+        else if(this->matched_begin && ! this->matched_end && matching)
+        {
+            if constexpr (Stoppable<TimerType>) {
+                this->timer.stop();
+            } else if constexpr (StoppableWithEvent<TimerType, EventType>) {
+                this->timer.stop(event);
+            }
+
+            this->matched_end = true;
```

Comment:
For the symmetry with the begin case.
```suggestion
            this->matched_end = true;
   
            if constexpr (Stoppable<TimerType>) {
                this->timer.stop();
            } else if constexpr (StoppableWithEvent<TimerType, EventType>) {
                this->timer.stop(event);
            }
```